src/phoenix/trace/otel/utils.py

```diff
@@ -0,0 +1,372 @@
+import json
+from collections import defaultdict
+from datetime import datetime, timezone
+from types import MappingProxyType
+from typing import (
+    Any,
+    DefaultDict,
+    Dict,
+    Iterable,
+    Iterator,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    Tuple,
+    cast,
+)
+from uuid import UUID
+
+import opentelemetry.proto.trace.v1.trace_pb2 as otlp
+from opentelemetry.proto.common.v1.common_pb2 import AnyValue, ArrayValue, KeyValue
+from opentelemetry.util.types import Attributes, AttributeValue
+from typing_extensions import TypeAlias, assert_never
+
+from phoenix.trace.schemas import (
+    Span,
+    SpanContext,
+    SpanEvent,
+    SpanException,
+    SpanID,
+    SpanKind,
+    SpanStatusCode,
+    TraceID,
+)
+from phoenix.trace.semantic_conventions import (
+    DOCUMENT_METADATA,
+    EMBEDDING_EMBEDDINGS,
+    EXCEPTION_ESCAPED,
+    EXCEPTION_MESSAGE,
+    EXCEPTION_STACKTRACE,
+    EXCEPTION_TYPE,
+    LLM_INPUT_MESSAGES,
+    LLM_OUTPUT_MESSAGES,
+    LLM_PROMPT_TEMPLATE_VARIABLES,
+    OPENINFERENCE_SPAN_KIND,
+    RETRIEVAL_DOCUMENTS,
+)
+
+
+def decode(otlp_span: otlp.Span) -> Span:
+    trace_id = cast(TraceID, _decode_identifier(otlp_span.trace_id))
+    span_id = cast(SpanID, _decode_identifier(otlp_span.span_id))
+    parent_id = _decode_identifier(otlp_span.parent_span_id)
+
+    start_time = _decode_unix_nano(otlp_span.start_time_unix_nano)
+    end_time = (
+        _decode_unix_nano(otlp_span.end_time_unix_nano) if otlp_span.end_time_unix_nano else None
+    )
+
+    attributes = dict(_decode_key_values(otlp_span.attributes))
+    span_kind = SpanKind(attributes.pop(OPENINFERENCE_SPAN_KIND, None))
+
+    for prefix in (
+        RETRIEVAL_DOCUMENTS,
+        EMBEDDING_EMBEDDINGS,
+        LLM_INPUT_MESSAGES,
+        LLM_OUTPUT_MESSAGES,
+    ):
+        consolidated_list, consolidated_keys = _consolidate_prefixed_indexed_keys_into_list(
+            attributes, prefix
+        )
+        if not consolidated_keys:
+            continue
+        for key in consolidated_keys:
+            attributes.pop(key, None)
+        if consolidated_list:
+            attributes[prefix] = consolidated_list
+
+    for document in attributes.get(RETRIEVAL_DOCUMENTS) or ():
+        if isinstance(metadata := document.get(DOCUMENT_METADATA), str):
+            try:
+                document[DOCUMENT_METADATA] = json.loads(metadata)
+            except json.JSONDecodeError:
+                pass
+
+    for prefix in (LLM_PROMPT_TEMPLATE_VARIABLES,):
+        consolidated_dict, consolidated_keys = _consolidate_prefixed_keys_into_dict(
```

Comment:
probably not a problem but the variable shadowing from above did trip me up a tad